github_pipeline_observability/pipeline_observability/document-dashboard/github.workflow.json

```diff
@@ -0,0 +1,3532 @@
+        },
+        "histogram": {
+          "dataMappings": [],
+          "displayedFields": [],
+          "variant": "single"
+        },
+        "honeycomb": {
+          "colorMode": "color-palette",
+          "colorPalette": "blue",
+          "dataMappings": {
+            "value": "failureRate"
+          },
+          "displayedFields": [
+            "step.name",
+            "step.task.name"
+          ],
+          "legend": {
+            "hidden": false,
+            "position": "auto"
+          },
+          "shape": "hexagon"
+        },
+        "icon": {
+          "icon": "",
+          "showIcon": false
+        },
+        "label": {
+          "showLabel": false
+        },
+        "singleValue": {
+          "alignment": "center",
+          "autoscale": true,
+          "colorThresholdTarget": "value",
+          "label": "error",
+          "prefixIcon": "",
+          "recordField": "error",
+          "showLabel": true,
+          "trend": {
+            "isVisible": true,
+            "trendType": "auto"
+          }
+        },
+        "table": {
+          "columnTypeOverrides": [],
+          "columnWidths": {},
+          "enableSparklines": false,
+          "hiddenColumns": [],
+          "lineWrapIds": [],
+          "linewrapEnabled": false,
+          "monospacedFontColumns": [],
+          "monospacedFontEnabled": false,
+          "rowDensity": "condensed"
+        },
+        "thresholds": [],
+        "tooltip": {
+          "showCustomFields": false
+        },
+        "valueBoundaries": {
+          "max": "auto",
+          "min": "auto"
+        }
+      }
+    },
+    "15": {
+      "davis": {
+        "davisVisualization": {
+          "isAvailable": true
+        },
+        "enabled": false
+      },
+      "query": "fetch events \n| filter event.kind == \"SDLC_EVENT\" \n| filter event.category == \"task\" \n| filter event.type== \"build\"\n| filter event.status == \"finished\" \n| filter event.provider == \"github.com\" \n| join [\n    fetch events \n    | filter event.kind == \"SDLC_EVENT\" \n    | filter event.category == \"pipeline\"\n    | filter event.type == \"run\" \n    | filter event.status == \"finished\" \n    | fields cicd.pipeline.run.id, cicd.pipeline.name, vcs.repository.name, cicd.pipeline.run.trigger \n    | filter in(vcs.repository.name, $Repository) \n    | filter in(cicd.pipeline.name, $Workflow) \n    | filter in(cicd.pipeline.run.trigger, $Trigger) ], \n    on:{cicd.pipeline.run.id}\n| fieldsAdd steps_array=toArray(task.steps) \n| expand steps_array \n| parse steps_array,\"JSON:steps_data\" \n| fieldsAdd step.duration = toTimestamp(steps_data[completed_at]) - toTimestamp(steps_data[started_at]), \nstep.name = steps_data[name], \nstep.task.name = task.name\n| summarize percentile(step.duration,50), alias:Median ,by:{step.name, alias: StepName, step.task.name, alias: TaskName} \n| filter isNotNull(Median) \n| sort Median desc \n| limit 5",
```

Comment:
Similar to above. Check if `summarize` is an option, instead of joining events.

Reply:
This query does not need a `join`: 
```
fetch events 
| filter event.kind == "SDLC_EVENT" 
| filter (event.category == "task" and event.status == "finished" and event.provider == "github.com") 
  or
  ( event.category == "pipeline" and event.type == "run" and event.status == "finished" 
  )
| summarize { 
  tasks = arrayRemoveNulls(collectArray(if(event.category == "task", record(task.name, duration), else: null))),
  has_matching_runs = countif(event.category == "pipeline")
}, by: {
  cicd.pipeline.run.id
}
| expand tasks
| filterOut has_matching_runs == 0 
| summarize {
   duration = percentile(tasks[duration], 50)
}, by: {
  task.name = tasks[task.name]
}
| sort duration desc 
| limit 5
```